### PR TITLE
Fix pagination when returning 0 objects

### PR DESCRIPTION
**Pull request by hlloreda**
# Description

Fixed metadata + link when returning 0 objects

Related to https://github.com/inmanta/inmanta-core/issues/7898

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/data/dataview.py [hunk 1]
```diff
@@ -139,6 +155,21 @@ def has_start(self) -> bool:
...
 
+    def get_empty_page_boundaries(self) -> PagingBoundaries:
+        """
+        Return the virtual paging boundaries that corresponds to the boundaries of an empty page response for this request.
+
+        Concretely, this means swapping start and end, because the semantics of PagingBoundaries are opposite those of this
+        class: PagingBoundaries represents "current page" while this class represents "requested page";
+        max of current page = min of requested page and vice versa.
+        """
+        return PagingBoundaries(
```
Comment (sanderr) [inline src/inmanta/data/dataview.py:166]: Perhaps we should add a comment here noting that there is a potential off-by-one error here because both classes are exclusive of the boundary values, and add a reference to the remaining ticket?

File: tests/server/test_agents_api.py [hunk 2]
```diff
@@ -216,6 +216,22 @@ async def test_agents_paging(server, client, env_with_agents: None, environment:
 
     assert result.result["metadata"] == {"total": 7, "before": 0, "after": 0, "page_size": 100}
 
+    fake_id_url = f"{base_url}/api/v2/agents?limit=1&sort=last_failover.asc&filter.status=paused&first_id=zzzz"
```
Comment (sanderr) [inline tests/server/test_agents_api.py:219]:
```suggestion
    # verify that paging beyond the last page returns correct counts and links
    fake_id_url = f"{base_url}/api/v2/agents?limit=1&sort=last_failover.asc&filter.status=paused&first_id=zzzz"
```

File: tests/server/test_agents_api.py [hunk 3]
```diff
@@ -216,6 +216,22 @@ async def test_agents_paging(server, client, env_with_agents: None, environment:
...
+    )
+    response = await http_client.fetch(request, raise_error=False)
+    assert response.code == 200
+    response = json.loads(response.body.decode("utf-8"))
+    # We don't have a way to reconstruct the previous link
+    assert response["links"] == {
+        "first": "/api/v2/agents?limit=1&sort=last_failover.asc&filter.status=paused",
+        "prev": "/api/v2/agents?limit=1&sort=last_failover.asc&filter.status=paused&last_id=zzzz",
+    }
+    assert response["metadata"] == {"total": 2, "before": 2, "after": 0, "page_size": 1}
```
Comment (sanderr) [inline tests/server/test_agents_api.py:232]: What does `page_size` represent? Should it be 1?

File: tests/server/test_agents_api.py [hunk 4]
```diff
@@ -216,6 +216,22 @@ async def test_agents_paging(server, client, env_with_agents: None, environment:
...
 
+    fake_id_url = f"{base_url}/api/v2/agents?limit=1&sort=last_failover.asc&filter.status=paused&first_id=zzzz"
+    request = HTTPRequest(
+        url=fake_id_url,
+        headers={"X-Inmanta-tid": str(environment)},
+    )
+    response = await http_client.fetch(request, raise_error=False)
+    assert response.code == 200
+    response = json.loads(response.body.decode("utf-8"))
+    # We don't have a way to reconstruct the previous link
```
Comment (sanderr) [inline tests/server/test_agents_api.py:227]: How do you mean? We do get a valid prev link, don't we?

File: tests/server/test_resource_list.py [hunk 5]
```diff
@@ -449,6 +450,43 @@ async def test_resources_paging(server, client, order_by_column, order, env_with
     assert response["metadata"] == {"total": 5, "before": 2, "after": 1, "page_size": 2}
 
 
+async def test_none_resources_paging(server, client, env_with_resources):
+    """Test querying resources with paging, using different sorting parameters."""
```
Comment (sanderr) [inline tests/server/test_resource_list.py:454]: This docstring doesn't seem to apply to what's being tested.

File: tests/server/test_resource_list.py [hunk 6]
```diff
@@ -449,6 +450,43 @@ async def test_resources_paging(server, client, order_by_column, order, env_with
...
+    env = env_with_resources
+    order_by_column = "agent"
+    order = "DESC"
+
+    no_result = await client.resource_list(
+        env.id,
+        limit=2,
+        sort=f"{order_by_column}.{order}",
+        start="ausf",
+        first_id="anet_co%3A%3Aresource%3A%3AResource%5BINMANTA-INMANTA-INMT%2Cresource_id%3Dinmt%5D",
```
Comment (sanderr) [inline tests/server/test_resource_list.py:464]: Apart from the confidentiality thing, which has been resolved, is there any reason we use so complex (difficult to read) resource here, rather than something simple and obviously meaningless like you have in the other test case with "zzz"?

File: tests/server/test_resource_list.py
Comment (sanderr) [inline tests/server/test_resource_list.py:1]: Can you add some scenarios for the other sorting orders: DESC-prev, ASC-next, ASC-prev?

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/data/model.py [hunk 7]
```diff
@@ -459,10 +459,22 @@ class PagingBoundaries:
...
+    neighbouring pages. Which represents next and which prev depends on sorting order (ASC or DESC).
+
+    Boundary values represent max and min values, regardless of sorting direction (ASC or DESC), i.e.
+    - ASC sorting based on next links: (>) end,last_id
+    - ASC sorting based on prev links: (<) start,first_id
+    - DESC sorting based on next links: (<) start,first_id
+    - DESC sorting based on prev links: (>) end,last_id
+
+    So, while the names "start" and "end" might seem to indicate "left" and "right" of the page, they actually mean "highest"
+    and "lowest".
```
Comment (Hugo-Inmanta) [inline src/inmanta/data/model.py:472]: While trying to understand this I added a tangible example with 2 columns (int, letter) that shows which param is taken into consideration on which page depending on the sorting order. It also shows the inclusive/exclusive ie what is on which page. Is this accurate ? (Feel free to ignore)
```suggestion
    and "lowest".
    
                                                                                                                                     
                  |        prev             |                current page                    |    next                           
     -------------|-------------------------|------------------------------------------------|----------------------             
     ASC sorting  | [  ...  ] (100 c)       | [ (100 c) ( 100 d) ... (10 000 r) (10 000 s) ] | (10 000 s) [ ...   ]              
                  |            start = 100  |     end = 100            start = 10 000        |     end = 10 000                  
                  |            first_id = c |     last_id = c          first_id = s          |     last_id = s                   
                  |                                                                                                              
     DESC sorting | [  ...  ] (10 000 s)    | [ (10 000 s) (10 000 r) ... (100 d) ( 100 c) ] | (100 c) [ ...   ]                 
                  |            end = 10 000 |          start = 10000          end = 100      |     start = 100                   
                  |            last_id = s  |          first_id = s           last_id = c    |     first_id = c                  
                                                                                                                                 
```

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: src/inmanta/data/model.py (hunk 7 above)
Comment (hlloreda) [reply src/inmanta/data/model.py:477]: I adjusted some values, but it's way cleaner, thank you!

File: tests/server/test_agents_api.py (hunk 3 above)
Comment (hlloreda) [reply tests/server/test_agents_api.py:233]: It's the number of objects that we asked for. I rewrote that part to make it clearer

File: src/inmanta/data/dataview.py [hunk 8]
```diff
@@ -96,7 +96,28 @@
...
+    Boundary values represent min and max values, regardless of sorting direction (ASC or DESC), i.e.
+                  |        prev             |                current page                    |    next
+     -------------|-------------------------|------------------------------------------------|----------------------
+     ASC sorting  | [  ...  ] (99 c)        | [ (100 d) ( 100 e) ... (10 000 r) (10 000 s) ] | (10 001 t) [ ...   ]
+                  |            end = 99     |     start = 100          end = 10 000          |     start = 10 001
+                  |            last_id = c  |     first_id = d         last_id = s           |     last_id = t
+                  |
+     DESC sorting | [  ...  ] (9 999 s)     | [ (10 000 t) (10 000 r) ... (100 d) ( 100 c) ] |  (99 b) [ ...   ]
+                  |          start = 9 999  |         end = 10000          start = 100       |     end = 99
+                  |          first_id = s   |         last_id = t          first_id = c      |     last_id = b
```
Comment (Hugo-Inmanta) [inline src/inmanta/data/dataview.py:111]: I think ?
```suggestion
     DESC sorting | [  ...  ] (10 001 t)    | [ (10 000 s) (10 000 r) ... (100 d) ( 100 c) ] |  (99 b) [ ...   ]
                  |          start = 10 001 |         end = 10000          start = 100       |     end = 99
                  |          first_id = t   |         last_id = s          first_id = c      |     last_id = b
```

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: tests/server/test_resource_list.py [hunk 9]
```diff
@@ -449,6 +450,136 @@ async def test_resources_paging(server, client, order_by_column, order, env_with
...
+    assert actual_result_desc_prev.result["links"] == expected_result_desc_prev
+    assert actual_result_desc_prev.result["metadata"] == {"after": 0, "before": 4, "page_size": 2, "total": 6}
+    assert len(actual_result_desc_prev.result["data"]) == 2
+
+    no_result_desc_next = await client.resource_list(
+        env.id,
+        limit=2,
+        sort="agent.DESC",
+        start="ausf",
+        first_id="anet_co",
```
[the remark this answers is not in the document]
Comment (hlloreda) [reply tests/server/test_resource_list.py:495]: We cannot do this because we want to have a page boundary without a next in with a descending order.

[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply tests/server/test_resource_list.py:495]: >  without a next

Without a prev you mean right?

[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply tests/server/test_resource_list.py:495]: I would expect in desc order `start="zzzz", first_id="zz"` to have the bulk of the data to the right and nothing as prev

File: src/inmanta/data/dataview.py [hunk 10]
```diff
@@ -96,7 +96,23 @@
...
 class RequestedPagingBoundaries:
-    """Represents the lower and upper bounds that the user requested for the paging boundaries, if any."""
+    """
+    Represents the lower and upper bounds that the user requested for the paging boundaries, if any.
+
+    Boundary values represent max and min values, regardless of sorting direction (ASC or DESC), i.e.
+    - ASC sorting based on next links: (>) start,first_id
+    - ASC sorting based on prev links: (<) end,last_id
+    - DESC sorting based on next links: (<) end,last_id
+    - DESC sorting based on prev links: (>) start,first_id
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/data/dataview.py]: My bad. gt, lt, lt, gt. I found it difficult to put in words. It's supposed to clarify the direction of the bounds, e.g. the first one means a page is requested for which all elements are `> (start, first_id)`.

File: changelogs/unreleased/7898-fix-pagination.yml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+description: "Fixed pagination when on a page returning 0 results (due to pagination criteria) that is strictly after or strictly before the bulk of the actual results"
```
Comment (hlloreda) [inline changelogs/unreleased/7898-fix-pagination.yml:1]: behind the last result  / before first result

File: src/inmanta/data/dataview.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hlloreda) [reply src/inmanta/data/dataview.py]: Add this to the docstring

File: src/inmanta/data/dataview.py [hunk 12]
```diff
@@ -400,48 +434,45 @@ async def prepare_paging_links(
...
+        if meta.before > 0:
+            if self.order.get_order() == "DESC":
+                links["prev"] = link_with_start
+            else:
+                links["prev"] = link_with_end
+            # First page
+            links["first"] = make_link()
+
+        # Same page
+        if dtos:
```
Comment (hlloreda) [inline src/inmanta/data/dataview.py:475]: api could assume self entry even if page empty

File: src/inmanta/data/dataview.py [hunk 13]
```diff
@@ -98,7 +98,28 @@
...
+    Boundary values represent min and max values, regardless of sorting direction (ASC or DESC), i.e.
+                  |        prev             |                current page                    |    next
+     -------------|-------------------------|------------------------------------------------|----------------------
+     ASC sorting  | [  ...  ] (99 c)        | [ (100 d) ( 100 e) ... (10 000 r) (10 000 s) ] | (10 001 t) [ ...   ]
+                  |            end = 99     |     start = 100          end = 10 000          |     start = 10 001
+                  |            last_id = c  |     first_id = d         last_id = s           |     first_id = t
+                  |
+     DESC sorting | [  ...  ] (10 001 t)    | [ (10 000 s) (10 000 r) ... (100 d) ( 100 c) ] |  (99 b) [ ...   ]
+                  |          start = 10 001 |         end = 10000          start = 100       |     end = 99
+                  |          first_id = t   |         last_id = s          first_id = c      |     last_id = b
```
Comment (Hugo-Inmanta) [inline src/inmanta/data/dataview.py:113]: Same here, I'm not so sure of what we're conveying here, maybe adds more confusion than anything

File: src/inmanta/data/model.py [hunk 14]
```diff
@@ -471,10 +471,27 @@ class PagingBoundaries:
...
+    Boundary values represent max and min values, regardless of sorting direction (ASC or DESC), i.e.
+                  |        prev             |                current page                    |    next
+     -------------|-------------------------|------------------------------------------------|----------------------
+     ASC sorting  | [  ...  ] (99 c)        | [ (100 d) ( 100 e) ... (10 000 r) (10 000 s) ] | (10 001 t) [ ...   ]
+                  |            start = 99   |     end = 100            start = 10 000        |     end = 10 001
+                  |            first_id = c |    last_id = d          first_id = s           |     first_id = t
+                  |
+     DESC sorting | [  ...  ] (10 001 s)    | [ (10 000 t) (10 000 r) ... (100 d) ( 100 c) ] |  (99 b) [ ...   ]
+                  |            end = 10 001 |          start = 10000          end = 100      |     start = 99
+                  |            last_id = t  |          first_id = t           last_id = c    |     first_id = b
```
Comment (Hugo-Inmanta) [inline src/inmanta/data/model.py:486]: I'm not sure I understand what's conveyed here anymore. We should remove it if it adds more confusion than anything.
Alternatively we could maybe add a single scenario where we describe what happens in the simple case eg:

User requests all Resources with name  >= foo in ASC order. page size = 10
```
RequestPagingBoundary:
    start = foo
    end = None
```
    
Actual data = [foo1 ... foo10]

```
Pagingboundary:
    end = foo1
    start = foo10    # Reversed because these are meant to be used by the neighbouring pages
```
    
Does this make sense / is actually correct ?

Comment (hlloreda) [reply src/inmanta/data/model.py]: Yes, well explained, thank you!

**Review by Hugo-Inmanta — APPROVED** [review]

File: src/inmanta/data/dataview.py [hunk 15]
```diff
@@ -141,6 +157,24 @@ def has_start(self) -> bool:
     def has_end(self) -> bool:
         return (self.end is not None) or (self.last_id is not None)
 
+    def get_empty_page_boundaries(self) -> PagingBoundaries:
+        """
+        Return the virtual paging boundaries that corresponds to the boundaries of an empty page response for this request.
+
+        Concretely, this means swapping start and end, because the semantics of PagingBoundaries are opposite those of this
+        class: PagingBoundaries represents "current page" while this class represents "requested page";
+        max of current page = min of requested page and vice versa.
```
Comment (sanderr) [inline src/inmanta/data/dataview.py:166]:
```suggestion
        class: PagingBoundaries' boundary values map to RequestedPagingBoundaries` boundary values for the next and previous
        pages, e.g. requested max = requested min of next page and vice versa.
```

File: src/inmanta/data/model.py [hunk 16]
```diff
@@ -473,16 +473,30 @@ class PagingBoundaries:
...
-    - ASC sorting based on next links: (>) end,last_id
-    - ASC sorting based on prev links: (<) start,first_id
-    - DESC sorting based on next links: (<) start,first_id
-    - DESC sorting based on prev links: (>) end,last_id
-
     So, while the names "start" and "end" might seem to indicate "left" and "right" of the page, they actually mean "highest"
     and "lowest".
 
+    Let's show this in an example: a user requests the following:
+     - all Resources with name >= foo
```
Comment (sanderr) [inline src/inmanta/data/model.py:480]:
```suggestion
     - all Resources with name > foo
```

File: src/inmanta/data/model.py [hunk 17]
```diff
@@ -473,16 +473,30 @@ class PagingBoundaries:
...
+            end = None
+        ```
+
+    The fetched data will be: [foo1 ... foo10]
+
+    But the Pagingboundary will be constructed this way:
+        ```
+        Pagingboundary:
+            end = foo1
+            start = foo10    # Reversed because these are meant to be used by the neighbouring pages
```
Comment (sanderr) [inline src/inmanta/data/model.py:497]:
```suggestion
            start = foo10 # Reversed because these are meant to map to like-named fields on neighbouring RequestedPagingBoundary
```

**Review by sanderr — CHANGES_REQUESTED** [review]

**Comment by sanderr** [comment]
@Hugo-Inmanta I made some final small changes to try to clarify so I'm rerequesting your review.

**Review by Hugo-Inmanta — APPROVED** [review]

**Review by sanderr — APPROVED** [review]
